contracts/contracts/MACI.sol

```diff
@@ -58,6 +60,15 @@ contract MACI is IMACI, DomainObjs, Params, Utilities {
   /// balance per user
   InitialVoiceCreditProxy public immutable initialVoiceCreditProxy;
 
+  /// @notice The array of the state tree roots for each sign up
+  /// For the N'th sign up, the state tree root will be stored at the index N
+  uint256[] public stateRootsOnSignUp;
+
+  // TODO: they are declared in other files as immutable
+  IVerifier private verifier;
```

Comment:
Both are used in Poll contract for verification proof

Reply:
so you can send them to the poll as ExtContracts no need to store them in the MACI contract. Same as mode, not sure why that is needed here?

Reply:
The problem is with the solidity limit of the arguments in a function call, because we need to forward verifier, vkRegystry and mode... Compiling the contract will fail

Reply:
you can pass a struct as parameter with all contract to `deployPoll` so you can group all of them in one

Reply:
https://github.com/0x3327/maci/commit/e70fd5bc03e64aa47d7b393f5e840194d38c7139

- removed verifier and vkRegistry from MACI
- add verifier and vkRegistry in Poll
- added `viaIR:true` in hardhat.config for compilation: https://github.com/0x3327/maci/blob/e70fd5bc03e64aa47d7b393f5e840194d38c7139/contracts/hardhat.config.ts#L56

Please review whether it is done correctly

Reply:
Good attempt, but there is no need to use viaIR. You can pass the addresses as struct to the pollFactory

MACI:

```solidity
    ExtContracts memory extContracts = ExtContracts({
      maci: IMACI(address(this)),
      verifier: IVerifier(_verifier),
      vkRegistry: IVkRegistry(_vkRegistry)
    });
    
    // the owner of the message processor and tally contract will be the msg.sender
    address _msgSender = msg.sender;

    address p = pollFactory.deploy(
      _duration,
      maxVoteOptions,
      _treeDepths,
      _messageBatchSize,
      _coordinatorPubKey,
      extContracts
    );
```

PollFactory

```solidity

  /// @inheritdoc IPollFactory
  function deploy(
    uint256 _duration,
    uint256 _maxVoteOptions,
    TreeDepths calldata _treeDepths,
    uint8 _messageBatchSize,
    PubKey calldata _coordinatorPubKey,
    ExtContracts calldata _extContracts
  ) public virtual returns (address pollAddr) {
    /// @notice Validate _maxVoteOptions
    /// maxVoteOptions must be less than 2 ** 50 due to circuit limitations;
    /// it will be packed as a 50-bit value along with other values as one
    /// of the inputs (aka packedVal)
    if (_maxVoteOptions >= (2 ** 50)) {
      revert InvalidMaxVoteOptions();
    }

    // deploy the poll
    Poll poll = new Poll(
      _duration,
      _maxVoteOptions,
      _treeDepths,
      _messageBatchSize,
      _coordinatorPubKey,
      _extContracts
    );
```

Poll

```solidity
 constructor(
    uint256 _duration,
    uint256 _maxVoteOptions,
    TreeDepths memory _treeDepths,
    uint8 _messageBatchSize,
    PubKey memory _coordinatorPubKey,
    ExtContracts memory _extContracts
  ) payable {
```

Reply:
https://github.com/0x3327/maci/commit/74a2ea1d67597cb48b8aedefd70676704bbd5019

Reply:
thank you!